File: src/main/java/brightspark/runicmagic/init/RMBlocks.java
package brightspark.runicmagic.init;

import brightspark.runicmagic.RunicMagic;
import brightspark.runicmagic.block.BlockGatestone;
import brightspark.runicmagic.block.BlockObelisk;
import brightspark.runicmagic.block.BlockRuneOre;
import brightspark.runicmagic.block.tileentity.TileGatestone;
import brightspark.runicmagic.enums.RuneType;
import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.fml.common.registry.GameRegistry;

import java.util.LinkedList;
import java.util.List;

@GameRegistry.ObjectHolder(RunicMagic.MOD_ID)
public class RMBlocks
{
	private static List<Item> ITEMS = new LinkedList<>();
	private static List<Block> BLOCKS = new LinkedList<>();

	public static final Block rune_ore = null;

	public static final Block obelisk_water = null;
	public static final Block obelisk_earth = null;
	public static final Block obelisk_fire = null;
	public static final Block obelisk_air = null;

	public static final Block gatestone = null;

	private static void add(Block block, Item itemBlock)
	{
		BLOCKS.add(block);
		ITEMS.add(itemBlock.setRegistryName(block.getRegistryName()));
	}

	private static void add(Block block)
	{
		add(block, new ItemBlock(block));
	}

	private static void regTE(Class<? extends TileEntity> tileEntityClass, Block block)
	{
		GameRegistry.registerTileEntity(tileEntityClass, block.getRegistryName());
	}

	private static void init()
	{
		add(new BlockRuneOre());

		add(new BlockObelisk(RuneType.WATER));
		add(new BlockObelisk(RuneType.EARTH));
		add(new BlockObelisk(RuneType.FIRE));
		add(new BlockObelisk(RuneType.AIR));

		Block block;
		add(block = new BlockGatestone(), new ItemBlock(block).setMaxStackSize(1));
		regTE(TileGatestone.class, block);
	}

	public static Block[] getBlocks()
	{
		if(BLOCKS.isEmpty())
			init();
		return BLOCKS.toArray(new Block[0]);
	}

	public static Item[] getItemBlocks()
	{
		if(ITEMS.isEmpty())
			init();
		return ITEMS.toArray(new Item[0]);
	}
}

File: src/main/kotlin/brightspark/runicmagic/init/RMBlocks.kt
package brightspark.runicmagic.init

import brightspark.runicmagic.block.GatestoneBlock
import brightspark.runicmagic.block.ObeliskBlock
import brightspark.runicmagic.block.RuneOreBlock
import brightspark.runicmagic.model.RuneType
import brightspark.runicmagic.util.setRegName
import net.minecraft.block.AbstractBlock.Properties
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.RenderTypeLookup
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.common.ToolType
import net.minecraftforge.event.RegistryEvent
import thedarkcolour.kotlinforforge.forge.objectHolder

object RMBlocks {
	val BLOCKS: MutableList<Block> = mutableListOf()

	val GATESTONE: Block by objectHolder("gatestone")

	fun register(event: RegistryEvent.Register<Block>) = event.registry.registerAll(
		runeOreBlock("rune_ore"),
		*RuneType.ELEMENTAL_TYPES.map { obeliskBlock("obelisk_$it", it) }.toTypedArray(),
		gatestoneBlock("gatestone")
	)

	@OnlyIn(Dist.CLIENT)
	fun clientInit() {
		RenderTypeLookup.setRenderLayer(GATESTONE, RenderType.getCutout())
	}

	private fun rockProps(
		material: Material = Material.ROCK,
		harvestLevel: Int = 0,
		hardness: Float = 3F,
		resistance: Float = 3F
	): Properties = Properties.create(material)
		.setRequiresTool()
		.harvestTool(ToolType.PICKAXE)
		.harvestLevel(harvestLevel)
		.hardnessAndResistance(hardness, resistance)

	private fun unbreakableProps(): Properties = Properties.create(Material.ROCK)
		.hardnessAndResistance(-1F, 3600000F)
		.noDrops()

	private fun Block.setup(name: String): Block = this.setRegName(name).apply { BLOCKS += this }

	private fun runeOreBlock(name: String): Block = RuneOreBlock(rockProps()).setup(name)

	private fun obeliskBlock(name: String, runeType: RuneType): Block =
		ObeliskBlock(runeType, unbreakableProps()).setup(name)

	private fun gatestoneBlock(name: String): Block =
		GatestoneBlock(
			Properties.create(Material.GLASS)
				.doesNotBlockMovement()
				.noDrops()
				.zeroHardnessAndResistance()
		).setup(name)
}
